Asiga001track volume

aswritescode opened the pull request.

aswritescode committed:
Added track_volume file

aswritescode committed:
Added basic printing statements

aswritescode committed:
Broke stuff

aswritescode committed:
Created simple test case and resolved compilation error

aswritescode committed:
Added more diagnostic prints

aswritescode committed:
Added some primitive logic for handling instructions

aswritescode committed:
Added primitive tracking functionality for Dispense and Dispose

aswritescode committed:
Added support for variables of size greater than 1. Fixed dispose logic

aswritescode committed:
Revised handle functions to accept IR object instead of IR.defs dict

aswritescode committed:
Added primitive mix tracking

aswritescode committed:
Fixed mix logic. Added internal dispose function

aswritescode committed:
Added more test cases. Added support for single-offset disposal

aswritescode committed:
Added primitve support for tracking split

aswritescode committed:
Fixed split issue. Added missing dispose call

aswritescode committed:
Added volume check to split

aswritescode committed:
Resolved issue with dispose function overloading

aswritescode committed:
Added volume check to dispose

aswritescode committed:
Adjusted tests

aswritescode committed:
Added support for custom volume capture

aswritescode committed:
Added support for Phi nodes

aswritescode committed:
Disabled some debug outputs

lilott8 commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -0,0 +1,252 @@
+    print("Handling phi...")
+
+    print("defs: " + str(instructions.defs))
+    print("uses: " + str(instructions.uses))
```
Please use self.log.[level] for printing.

lilott8 commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -0,0 +1,252 @@
+
+    violation_found = False
+
+    _program = None;
```
These belong in `__init__()`

unleavenedbread replied:
If I move the variable declarations to within init() then I can't access them later on in my code.

tysolov replied:
If they are not bound to self, then that is true; however, if you declare them, for example, as "self.variable_volume = dict()" and so on, then they are bound to the object instance.

lilott8 commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -0,0 +1,252 @@
+                        print(i)
+                    #print("i: " + str(i))
+                    #print("defs: " + str(i.defs))
+                    #print("uses: " + str(i.uses))
```
remove rem'd lines.

lilott8 commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -0,0 +1,252 @@
+
+
+def handle(self, instruction : IR): # The meat of the logic. Just ferries out the functions based on the type of instruction.
+    if (type(instruction) == Dispense):
```
IRInstruction has an `op` field that tells you what it is.  You don't need to do `type() == ...`, you can do `instruction.op == IR.[op]`

unleavenedbread replied:
Using Type() is actually very fast and functions well considering all Python variables are self-aware of their typing.

lilott8 commented on test_cases/dispense_if_else.bs:
```diff
@@ -0,0 +1,11 @@
+	a = dispense 10 units of fluid_a
+}
+
+b = split a into 5
```
All test cases belong in the `\tests\test_cases\volume\` folder.  You will then need to add a test harness to `\tests\frontend\` that handles volume tracking.  You can copy whatever of the boilerplate you need from other files.

unleavenedbread replied:
How do I access the global variables stored in my Pass's class? I don't see an obvious way of grabbing it using the existing functions in front_end_base.py

lilott8 reviewed (changes requested):
All test cases must be put in the \tests\...\... folder, whereby the pytest test harnesses can automatically run it.  All 100+ test cases should be passing, including your test cases, before another PR is issued.

aswritescode committed:
Removed debugging outputs

aswritescode committed:
Moved test cases and created test harness

aswritescode committed:
Added a test case

aswritescode committed:
Added extra files

aswritescode committed:
Merge branch 'master' into asiga001track_volume

aswritescode committed:
Added init files

aswritescode committed:
Fixed type error

aswritescode committed:
Removed -t parameter from launch in test harness

aswritescode committed:
Added dispense_if_else test case

aswritescode committed:
Added basic dispose test case

aswritescode committed:
Added several test cases

aswritescode committed:
Adjusted and implemented test cases

aswritescode committed:
Added missing test case

aswritescode committed:
Re-aranged problematic test's location

tysolov commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -0,0 +1,215 @@
+
+
+
+def handle(self, instruction : IR): # The meat of the logic. Just ferries out the functions based on the type of instruction.
```
This and the following functions should be declared within the class.

You'll need to call each method as `self.<method>`, and remove the `self` parameter from the call, as the calling class instance implicitly sends itself as the first parameter.

tysolov commented on compiler/semantics/ir_visitor.py:
```diff
@@ -755,10 +755,22 @@ def visitDispense(self, ctx: BSParser.DispenseContext):
+        else:
+            _volume = int(ctx.unitTracker().INTEGER_LITERAL().__str__())
+
+        #print("Volume: " + str(_volume)) #DEBUG
```
Remove dead code throughout.  For debugging, use the following pattern:
```
if self.config.debug:
   self.log.debug(<output>)
```

tysolov commented on tests/frontend/test_volume.py:
```diff
@@ -0,0 +1,148 @@
+
+        assert vol[0] == True
+
+    def test_if_else(self, get_visitor):
```
The result here does not seem correct.  There are two potential results that could come from the analysis on `dispense_if_else.bs`:

(1) if your analysis is capable of determining flow statically (as in this case 3<=3 is always true), then vol(a) = {5} prior to splitting, and null after the split, and each element of b should have b(i) = {1}.

(2) if your analysis ignores flow (minding not all conditionals can be statically determined, so this should be implemented either way when we have a variable conditional), then vol(a) = {5, 10} prior to split and null after split, and each element of b should have b(i) = {1, 2}.

tysolov reviewed (changes requested):
Address unresolved comments prior to next review request

aswritescode committed:
Removed dead code, converted static methods to class methods in track_volume.py

aswritescode committed:
Added primitive support for mix parameters to ir_visitor

aswritescode committed:
Added support for parameter passing for mix

aswritescode committed:
Updated track_volume.py to store all volumes states instead of only the most current one. It now passes this collection to the pass manager too. Updated test-harness to reflect this change

aswritescode committed:
Resolved issue with passing shalow copies of variable_volume to past_volumes

aswritescode committed:
Added support for phi statements to ir_visitor. Modified symbol to hold new structure

aswritescode committed:
Removed dead code

aswritescode committed:
Removed dead code

aswritescode committed:
Merge branch 'master' into asiga001track_volume

tysolov commented on compiler/semantics/ir_visitor.py:
```diff
@@ -757,17 +817,18 @@ def visitDispense(self, ctx: BSParser.DispenseContext):
 
-        #print("Scope_Stack: " + str(self.scope_stack[-1]))
+        if deff['name'] in self.symbol_table.get_local(deff['name'], self.scope_stack[-1]).volumes:
+            self.symbol_table.get_local(deff['name'], self.scope_stack[-1]).volumes[deff['name']].append(_volume)
```
Here you're using the variable name as the key to the volumes dictionary, which requires you to use some hacky manipulation to retrieve values in the volume tracker -- the whole point of using a dictionary is for uniqueness.  This works _for tested cases_, but as you are using the index of the list to retrieve the value based on the _assumed name_ from SSA-renaming, this does not generalize.  In short, the dict key should be the instruction, not the name, and when you retrieve a value in volume tracker you'll need that same line number.

You can get the _unique_ instruction id (`iid`) from the ir that is generated a few lines after this.

You will want to do the same for the mix value you parse here (unique dict items keyed by the `iid`), as there is nothing stopping us from doing:

```
a = dispense ....
a = mix x with y
```
Which would overwrite the volume attribute in the symbol table currently.

tysolov commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -26,190 +26,201 @@ def analyze(self, program: Program) -> dict: # The main function of the class
+        name_num_only = instructions.defs['name'][-1]
 
-    volumes.append(quantity) # Add the dispense'd quantity toi the volumes list. In this case we assume that size is always 1, so we only do it a single time.
+        quantity = self._program.symbol_table.get_local(instructions.defs['name'], "main").volumes[name_no_num][int(name_num_only) - 1]
```
The comment I left in `ir_visitor` refers to this hacky access.  Please update accordingly using `iid` or some other unique identifier.

tysolov commented on compiler/passes/analyses/track_volume.py:
```diff
@@ -26,190 +26,201 @@ def analyze(self, program: Program) -> dict: # The main function of the class
-        self.variable_volume[instructions['name']]['volumes'] = [-1]; # Since volumes is a list, we wrap our single volume data in its own list. This is to avoid any issues when reading a disposed variable's entry down the line
-        self.variable_volume[instructions['name']]['size']    = 0;    # A disposed variable doesn't have a presence on the board. It's size is therefore zero.
+    def handle_mix(self, instructions: IR):
+        encoded_quantity = self._program.symbol_table.get_local(instructions.defs['name'], "main").value.volume['quantity'] # The quantity stored in an encoded float value.
```
This uses the `value` attribute of a variable.  You should instead use your new `volumes` attribute, as `value` is updated to the _last parsed value_ with this variable name.

Also, consider the statement: `a = mix b with c`.

You do not know if `b` and `c` have only a single possible value, as they may be defined over various control paths; hence, you should have `volumes_b` and `volumes_c`, and iterate over the combination of each pair to form the set of possible volumes of `a`.

This logic should extend to split as well.

unleavenedbread replied:
I believe we discussed this and agreed that we could naively assume that a variable's volume is always equal to its lowest possible value.

tysolov reviewed (changes requested):
Please review updated comments and address accordingly.

aswritescode committed:
Replaced hackey encoding methods with proper iid identification

aswritescode committed:
Removed dead code

tysolov committed:
Various bug fixes:
 - UndefinedVariable now thrown prior to access in symbol_visitor.py
 - self.graph.add_cycle deprecated; updated with networkx.classes.function.add_cycle
 - z3_test.py is unfinished, erroring out in test suite, and has been marked as TODO
 - pytest marks registered in pytest.ini
